### ISSUE TITLE
[Sulong] Repeated printing of WARNING: setjmp is unsupported!

### ISSUE THREAD

**Issue by mrigger** [oracle/graal#1201]
Hi,

I'm executing a benchmark from the cicro benchmarks that calls `setjmp`, but never `longjmp`. Sulong does not support `setjmp` and repeatedly prints a warning `WARNING: setjmp is unsupported!` (see [here](https://github.com/oracle/graal/blob/7b0cc8ba7d6ce48d709e8e1f6551d06c4be1f2c9/sulong/projects/com.oracle.truffle.llvm.libraries.bitcode/src/setjmp.c#L36)). Only if `longjmp` is called would Sulong abort.

Would it be possible to either (a) remove the warning since Sulong would abort anyway when `longjmp` is called or (b) print the warning only once per execution?

**Comment by zapster** [comment]
Thanks Manuel for the report. IMHO It is tricky to deal with `setjmp`/`longjmp` in a graceful way.

Regarding you suggestion (a): I am concerned about the case when `setjmp` is called in bitcode (which would currently print the warning), but `longjmp` is called in native code, which means so we cannot intercept it. In that case, the program will abort but we would never notify the user that something unsupported is going on (especially what the problem is).

Regarding (b): Is that really much of an improvement? If processed in an automatic way (as in cicro) it will very likely fail anyway. If ran manually, it is easy to miss. So I am not convinced. :)

Long story short, I think I need to think more about it to come up with a proper solution.

**Comment by mrigger** [comment]
Thanks for your quick response! Regarding (b): It would be an improvement, for cicro at least, since for programs that call `setjmp` in a loop, large amounts of data on stderr are produced. This is a problem for cicro, since the `stderr` output is stored to a file and printed to the console.

Aborting the program when `setjmp` is called would also be satisfactory, which would prevent the case that `setjmp` is called from bitcode and `longjmp` from native code. The downside is that this specific benchmark (and probably other programs) would no longer run on Sulong.

**Comment by pekd** [comment]
I originally implemented these warnings/errors. Before `setjmp`/`longjmp` were implemented in bitcode, programs called the native implementation, which was no problem for `setjmp` but it caused weird and hard to understand JVM crashes for `longjmp`. The idea behind the warnings was that the user should know he is using something unsupported.

Aborting at `setjmp` is a bad idea, since many programs use `setjmp`/`longjmp` for error handling, and as long as no error occurs, ignoring the `setjmp` is perfectly valid. The main reason why the implementation of `setjmp` only prints a warning but doesn't abort is because I've seen so many programs using it without ever calling `longjmp`.

There was even a PR in the old Sulong repository for an experimental implementation of `setjmp`/`longjmp`, which worked as long as both functions are called from bitcode.

Printing the warning only once is probably the easiest and best "fix" for now?

**Comment by zapster** [comment]
Thanks @pekd for sharing your insights!

**Comment by zapster** [comment]
The warning has been removed in GraalVM 21.0.0.
